File: without-library/graph/traversal/bfs.py
import sys

# BFS without deque/collections
# Using list with pointer for O(1) avg pop

class BFS:
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.dist = [-1] * (n + 1)
        self.parent = [-1] * (n + 1)
        
    def bfs(self, start):
        # Manual queue using list
        queue = [start]
        head = 0 # Pointer to front
        
        self.dist[start] = 0
        self.parent[start] = -1
        
        while head < len(queue):
            u = queue[head]
            head += 1
            
            for v in self.adj[u]:
                if self.dist[v] == -1:
                    self.dist[v] = self.dist[u] + 1
                    self.parent[v] = u
                    queue.append(v)

if __name__ == "__main__":
    pass

File: without-library/graph/traversal/bfs.cpp
#include <iostream>
#include <vector>

using namespace std;

// BFS without std::queue

vector<int> adj[200005];
int dist[200005];
int parent[200005];

void bfs(int start, int n) {
    for(int i=0; i<=n; i++) dist[i] = -1;
    
    // Manual queue using vector
    vector<int> q;
    q.push_back(start);
    int head = 0;
    
    dist[start] = 0;
    parent[start] = -1;
    
    while(head < q.size()) {
        int u = q[head++];
        
        for(size_t i=0; i<adj[u].size(); i++) {
            int v = adj[u][i];
            if(dist[v] == -1) {
                dist[v] = dist[u] + 1;
                parent[v] = u;
                q.push_back(v);
            }
        }
    }
}

int main() {
    return 0;
}
